### PR TITLE
core(tracehouse): structure main thread task creation to be more forgiving

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
tl;dr - a rewrite of main-thread-tasks to make way for us dropping invalid trace events in the future

This PR converts our main thread creation to focus on individual tasks first, *then* moves on the hierarchy step. We were being fancy before by doing everything in a single pass, which is much more efficient and easier when the trace events are valid but next to impossible to understand if you're trying to recover from invalid data. By switching our approach, it's much easier to identify unmatched pairs of events, overextending tasks, etc.

~~It's *mostly* identical functionality-wise to our current setup with the exception of no longer requiring the logic in #9230 to handle same-timestamped events, so should be minimal impact as-is.~~ The matching of B/E events now necessitates a more forgiving parent/child event timing rule so we handle missorted events *and* we allow 1ms slop for parent/child start/end times.

The next step would be instead of `throw new Error` in main-thread-tasks it pushes the trace event onto an `invalidTraceEvents` array so that we can continue and the user still is informed there was something squirrely is going on.

**Related Issues/PRs**
moves us a step closer to addressing #7764

eliminates the need for #9230 in order to address #7764 though as discussed in the PR description, still worth doing IMO as it sets this code up to follow the hot path and is just a more sensible way to receive trace events.

### CODE REVIEW

**Comment by patrickhulce** [comment]
OK so turns out we already had an "invalid" trace checked into our fixtures and we just never noticed because the main thread task logic didn't check that the corresponding B/E event names matched, it just assumed they were valid. No we check them and so we started throwing `Child cannot end after parent errors`.

This PR now also allows up to 1ms slop time between parent/child events to allow for these minor mistakes and should hopefully reduce the number of fatal trace processing errors.

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 1]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-  static _createTasksFromEvents(mainThreadEvents, priorTaskData, traceEndTs) {
+  static _createTasksFromStartAndEndEvents(taskStartEvents, taskEndEvents, traceEndTs) {
     /** @type {TaskNode[]} */
     const tasks = [];
-    /** @type {TaskNode|undefined} */
-    let currentTask;
+    // Create a reversed copy of the array to avoid copying the rest of the queue on every mutation.
+    const taskEndEventsReverseQueue = taskEndEvents.slice().reverse();
+
+    for (let i = 0; i < taskStartEvents.length; i++) {
```
Comment (exterkamp) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: There are a lot of the nested loops where one side searches left to right, and then looks up right to left.  Is there any way to iterate and build a map and lookup the seen that way?  go from O(x^2) to O(x)? This might be overcomplicated for some code that is already complex.

^ That might not make 100% given this code is searching 2 separate lists and it is trying to handle nesting, but this definitely has a suspicious smell to the code. Might just need better comments.

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 2]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
       }
 
-      // Only look at X (Complete), B (Begin), and E (End) events as they have most data
-      if (event.ph !== 'X' && event.ph !== 'B' && event.ph !== 'E') continue;
+      tasks.push(MainThreadTasks._createNewTaskNode(taskStartEvent, taskEndEvent));
+    }
+
+    if (taskEndEventsReverseQueue.length) {
+      throw new Error(
+        `Fatal trace logic error - ${taskEndEventsReverseQueue.length} unmatched E events`
```
Comment (exterkamp) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: I assume `E` here is "End", maybe spell this out better since this is a semantic error message a user might see?

File: lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js [hunk 3]
```diff
@@ -147,11 +153,54 @@ describe('Main Thread Tasks', () => {
...
 
+  it('should compute attributableURLs correctly across timers', () => {
+    const baseTs = 1241250325;
+    const url = s => ({args: {data: {url: s}}});
+    const stackFrames = f => ({args: {data: {stackTrace: f.map(url => ({url}))}}});
+    const timerId = id => ({args: {data: {timerId: id}}});
+
+    /*
+    An artistic rendering of the below trace:
+    █████████████████████████████TaskA██████████████████████████████████████████████
```
Comment (exterkamp) [inline lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js]: I would love some of these renderings in the code as comments explaining what it is doing! :art:

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 4]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-      if (event.name === 'TimerInstall' && currentTask) {
-        /** @type {string} */
-        // @ts-ignore - timerId exists when name is TimerInstall
-        const timerId = event.args.data.timerId;
-        priorTaskData.timers.set(timerId, currentTask);
+      // Task is a B/E event pair, we need to find the matching E event.
+      let matchedEventIndex = -1;
+      let matchingNestedEventCount = 0;
+      let matchingNestedEventIndex = i + 1;
+      for (let j = taskEndEventsReverseQueue.length - 1; j >= 0; j--) {
```
Comment (paulirish) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: if we reversed this array it's unclear why we're iterating from the back to front.

this improves the perf of the pop() and splice()s?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 5]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+   *    2. Create tasks for each X/B event, throwing if a matching E event cannot be found for a given B.
+   *    3. Sort the tasks by ↑ startTime, ↓ duration.
+   *    4. Match each task to its parent, throwing if there is any invalid overlap between tasks.
+   *
+   * @param {LH.TraceEvent[]} mainThreadEvents
+   * @param {PriorTaskData} priorTaskData
+   * @param {number} traceEndTs
+   * @return {TaskNode[]}
+   */
+  static _createTasksFromEvents(mainThreadEvents, priorTaskData, traceEndTs) {
```
Comment (paulirish) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: in getMainThreadTasks() can you rename `traceEvents` to `mainThreadEvents`?

this grouping technique only works within a thread so i just want to be clear that we already assume pid/tid all match here.

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 6]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-        const timerId = event.args.data.timerId;
-        priorTaskData.timers.set(timerId, currentTask);
+      // Task is a B/E event pair, we need to find the matching E event.
+      let matchedEventIndex = -1;
+      let matchingNestedEventCount = 0;
+      let matchingNestedEventIndex = i + 1;
+      for (let j = taskEndEventsReverseQueue.length - 1; j >= 0; j--) {
+        const endEvent = taskEndEventsReverseQueue[j];
+        // We are considering an end event, so we'll count how many nested events we saw along the way.
+        while (matchingNestedEventIndex < taskStartEvents.length &&
```
Comment (paulirish) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: looking at https://cs.chromium.org/chromium/src/third_party/catapult/tracing/tracing/extras/importer/trace_event_importer.html (and [slice_group](https://cs.chromium.org/chromium/src/third_party/catapult/tracing/tracing/model/slice_group.html)) .. they use something very simple
basically since sync events are assumed to be sync... if we have an End, then it must match with the most recently found Begin event.

also interestingly they don't sort events or preprocess before doing this begin/end matching. perhaps making this assumption from the trace saves some headaches?

Comment (paulirish) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: i tried out a basic implementation of this approach but it fails on 

> should handle out-of-order 0 duration tasks
> should handle child events that extend <1ms beyond parent event

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 7]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
 
   /**
-   * @param {LH.TraceEvent[]} mainThreadEvents
+   *
+   * @param {TaskNode} currentTask
+   * @param {{startTime: number}} nextTask
    * @param {PriorTaskData} priorTaskData
+   * @param {Array<LH.TraceEvent>} reverseEventsQueue
+   */
+  static _assignAllTimerInstallsBetweenTasks(
```
Comment (paulirish) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: are there other test cases we should add for this new method?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > meanwhile, recordsFromLogs takes 1380ms.

Like network recordsFromLogs?? Even theverge will have max 100s of requests. What's taking 1.3s?

Comment (paulirish) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: findNetworkQuietPeriods

image

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 8]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+      currentTask,
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: I respectfully disagree, you just must not have seen very much nicely formatted JS ;)

All standalone functions with multiple lines in our codebase are formatted this way except 1, and that's only because it was [authored by an external contributor two years ago](https://github.com/GoogleChrome/lighthouse/commit/890103468bc724cc221e47bbbb91a21fb35efbe4#diff-e2079966cf5431d4d65b3c6dbd91b25cR37). This is also the formatting approach taken by [prettier](https://prettier.io/) which has become somewhat of a defacto standard style across new JS projects in the community.

I rest my case 👩‍⚖ 😄

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 9]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+        // We're currently in the middle of a task, so `nextTask` is a child.
+        nextTask.parent = currentTask;
+        currentTask.children.push(nextTask);
+        MainThreadTasks._assignAllTimerInstallsBetweenTasks(
+          currentTask,
+          nextTask,
+          priorTaskData,
+          timerInstallEventsReverseQueue
+        );
+      } else {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: Not sure how to reconcile this one with the request for more comments and documentation, this is explicitly here so that the reader knows what the else branch would have been and why it's not necessary

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 10]
```diff
@@ -44,104 +44,268 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
     }
+  }
 
-    // Starting from the last and bottom-most task, we finish any tasks that didn't end yet.
-    while (currentTask && !Number.isFinite(currentTask.endTime)) {
-      // The last event didn't finish before tracing stopped, use traceEnd timestamp instead.
-      currentTask.endTime = traceEndTs;
-      currentTask = currentTask.parent;
+  /**
+   * This function takes the raw trace events sorted in increasing timestamp order and outputs connected task nodes.
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: There is a top level description of this file that addresses all that this file does while this comment discusses how this piece of the puzzle is computed (the only parts of this file that were touched were the trace parsing pieces).

It sounds like you think it might be worth splitting this file into different components?

I'm not quite sure what that would look like though. Maybe just...

- main-thread-tasks.js
- thread-task-creation.js

?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: while it's true the worst case runtime here is O(n^2), ~99.999% of the time the reverse loop is O(1) because the common case is that B/E events are in the correct order (think of every time we've run LH today without it throwing a fatal error, that is all of the cases that this reverse loop was O(1) on every event). The only reason we're doing this at all is to be able to handle the rare and unexpected case where Chrome put the events in an order that doesn't make sense. When that happens we always have to look at every event worst case so there's no way to escape O(n^2).

Also, what paul said, it's fine right now so seems premature to worry about taking up extra memory with a map.

Seems like a good idea to put all this into a comment somewhere :)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: I'm slightly torn here because there are multiple letter designations for end events in general but `E` has a very explicit meaning for a trace event. 

If a user is getting this message, there's pretty much nothing they can do about it already, so I'm not sure helping them understand what an `E` event is better than providing debugging us with clarity on exactly which mismatch is happening. 

Then again all of these messages are pretty much useless on their own for debugging and we always need a full trace to investigate anyway so there's not much point in providing debug clarity in the message at all for us so long as they're unique, they could be random numbers and be just as useful for us.

I'm convinced. Sounds good 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: yep `pop() <<< shift()` added a comment to this effect and a better function explanation 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: yup sg 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > perhaps making this assumption from the trace saves some headaches?

It absolutely does save headaches and that's how the old version was written but that assumption is also what fails on several real-world traces. Recovering from that error situation is exactly what this rewrite is about, so if there's a way to get that old way to work and nicely recover from errors when that assumption breaks down I'm definitely open to someone else running with their vision for it :)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 7 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: I already added https://github.com/GoogleChrome/lighthouse/blob/6daf898a7f3a4257b21abc23597ffab10cefe63f/lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js#L156-L197 which is enough to cover the case I had in mind with this method but maybe there's another case I'm missing?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: yeah i prefer this too, but I disagree with any argument using "prettier does it" since I have grown to dislike how "prettier" code looks =P

[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > I disagree with any argument using "prettier does it" since I have grown to dislike how "prettier" code looks =P

That's fine but no matter how you feel about prettier it is clear evidence this style is present in the javascript community and should not feel like only cpp.

[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: motion to suppress evidence successful 🎉

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 11]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+    let currentTask;
+    // Create a reversed copy of the array to avoid copying the rest of the queue on every mutation.
+    const timerInstallEventsReverseQueue = timerInstallEvents.slice().reverse();
 
-      // Update currentTask based on the elapsed time.
-      // The next event may be after currentTask has ended.
+    for (let i = 0; i < sortedTasks.length; i++) {
+      const nextTask = sortedTasks[i];
+
+      // Update `currentTask` based on the elapsed time.
```
Comment (brendankenny) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: More appropriate below where timing is updated?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 12]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-        // We can't start a task with an end event
-        if (event.ph === 'E') {
-          throw new Error('Fatal trace logic error - unexpected end event');
+      if (currentTask) {
+        if (nextTask.endTime > currentTask.endTime) {
+          const timeDelta = nextTask.endTime - currentTask.endTime;
+          // The child task is taking longer than the parent task, which should be impossible.
+          //    If it's less than 1ms, we'll let it slide.
+          //    If it's more, throw an error.
+          if (timeDelta < 1000) {
```
Comment (brendankenny) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: is it weird to accept sloppiness in this direction but not in the other, 1ms starting too early? That could be 1ms before and then ending within `currentTask`, or what if starting within 1ms of `currentTask` ending but ending 100ms after (would currently throw but maybe should be considered non-nesting?)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 13]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+            currentTask.endTime = nextTask.endTime;
+            currentTask.duration += timeDelta;
```
Comment (brendankenny) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: this seems like something that will bite us when not accounted for elsewhere (you discussed this in #9230). Here be dragons, but maybe we should fix this up at the trace level...

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: FWIW, I think something like, `// If there's still a currentTask, that means we're currently in the middle of a task, so nextTask is a child.` up at the top of the `if` works (and then can drop the `else`)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 14]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
+      let matchingNestedEventIndex = i + 1;
+
+      // We loop through the reversed end events queue from back to front because we still want to
+      // see end events in increasing timestamp order.
+      // While worst case we will loop through all events, the overwhelmingly common case is that
+      // the immediate next event is our event of interest which makes this loop typically O(1).
+      for (let j = taskEndEventsReverseQueue.length - 1; j >= 0; j--) {
+        const endEvent = taskEndEventsReverseQueue[j];
+        // We are considering an end event, so we'll count how many nested events we saw along the way.
+        while (matchingNestedEventIndex < taskStartEvents.length &&
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: sg 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 15]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-    /** @type {TaskNode|undefined} */
-    let currentTask;
+    // Create a reversed copy of the array to avoid copying the rest of the queue on every mutation.
+    // i.e. pop() is O(1) while shift() is O(n), we take the earliest ts element off the queue *a lot*
+    // so we'll optimize for having the earliest timestamp events at the end of the array.
+    const taskEndEventsReverseQueue = taskEndEvents.slice().reverse();
+
+    for (let i = 0; i < taskStartEvents.length; i++) {
+      const taskStartEvent = taskStartEvents[i];
+      if (taskStartEvent.ph === 'X') {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: the solution to this isn't clear to me, so I might need some help :)

what's clear to me now is that every start task case is handled here which is comforting compared to making sure I have my bases covered by combining the results of several methods if that's what the suggestion is? IMO, it's maybe a wash considering the mental overhead of this is only 3 lines? might be missing what the separately altogether looks like though 🤷‍♂

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 16]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-      // Update currentTask based on the elapsed time.
-      // The next event may be after currentTask has ended.
+    for (let i = 0; i < sortedTasks.length; i++) {
+      const nextTask = sortedTasks[i];
+
+      // Update `currentTask` based on the elapsed time.
+      // The `nextTask` may be after currentTask has ended.
       while (
         currentTask &&
         Number.isFinite(currentTask.endTime) &&
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: Given the new construction method it should never be an issue when given a valid trace. Unfortunately I have no more confidence that this code will be given a valid trace these days 😢

[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: (to be fair though, it should be much much rarer than before where `dur` itself is not a finite number)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 12 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: the eventual goal is to accept sloppiness in both directions ya but it's much more complicated to accept it in the other direction so saving that for a separate PR.

TODO? issue?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 17]
```diff
@@ -44,104 +44,285 @@ const {taskGroups, taskNameToGroup} = require('./task-groups.js');
...
-        currentTask.endTime <= event.ts
+        currentTask.endTime <= nextTask.startTime
       ) {
+        MainThreadTasks._assignAllTimerInstallsBetweenTasks(
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: yeah sg, `_assignAllTimersUntilTs` is more accurate 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 11 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: I can see these two are being misread, I'll reword and add something different to below :)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 13 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: this is the only place where we consume nested tasks at all at the moment so we should be fine for now handling just here, but if the day comes where we need to then we can pair the output of this with a source trace to "fix" the trace events for consumption by other tools 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 12 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: did a TODO for now

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: yeah, it wouldn't be much different, just something like `if (event.ph === 'X') taskXEvents.push(event);` down in `_createTasksFromEvents` and a separate function `_createTasksFromXEvents` that's basically just this first three lines.

The main thing for me wasn't having this code here (though it would keep it slightly more focused), it was the `matchingNestedEventIndex` business below, where I was trying to figure out if the `name` check was sufficient to rule out `X` events, if having them in there would mess up the matching ordering in `taskEndEventsReverseQueue`, etc. All of that works fine, and there's no performance advantage, but it might be nice to separate them to ditch any mental overhead of having B and X together.

Not a huge deal, it also works as it is.

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 12 above)
Comment (brendankenny) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > TODO? issue?

yeah, either is fine. Just good to have it documented somewhere what we anticipate could be more sloppiness we have to account for. Helps the next person looking at a trace and saying "well that shouldn't be possible..." :)

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 13 above)
Comment (brendankenny) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > but if the day comes where we need to then we can pair the output of this with a source trace to "fix" the trace events for consumption by other tools 👍

SG

**Comment by brendankenny** [comment]
(with the really important lint fix)

**Comment by brendankenny** [comment]
haha gg eslint
